File: tailwind-admin-reactjs-free/package/src/types/calificaciones.ts
export interface EvaluacionGrupoMini {
  id_grupo: number;
  name_grupo: string;
}

export interface EvaluacionAsignacion {
  id_profesor_asignatura_grupo: number;
  profesor: { id_profesor: number; name_profesor: string; sec_name_profesor: string };
  asignatura: { id_asignatura: number; name_asignatura: string };
}

export type TipoRubro = 'Examen' | 'Tarea' | 'Cotidiano';

export interface Evaluacion {
  id_evaluacion: number;
  name_evaluacion: string;
  tipo: TipoRubro;
  periodo: number;
  porcentaje: number;
  fecha: string | null;
  id_grupo: number;
  activo: boolean;
  grupo: EvaluacionGrupoMini;
  id_profesor_asignatura_grupo: number | null;
  asignacion: EvaluacionAsignacion | null;
}

export interface EvaluacionInput {
  name_evaluacion: string;
  tipo: TipoRubro;
  periodo: number;
  porcentaje: number;
  fecha: string | null;
  /** Nuevo modelo: la evaluacion cuelga de una asignacion profesor+materia. */
  id_profesor_asignatura_grupo?: number;
  /** Compatibilidad: crear directamente por grupo (sin asignacion). */
  id_grupo?: number;
}

export interface NotaRosterItem {
  id_estudiante: number;
  name_estudiante: string;
  sec_name_estudiante: string;
  valor: number | null;
}

export interface NotasRoster {
  id_evaluacion: number;
  name_evaluacion: string;
  registros: NotaRosterItem[];
}

export interface NotaRegistroIn {
  id_estudiante: number;
  valor: number | null;
}

export interface NotasBatchIn {
  registros: NotaRegistroIn[];
}

File: backend/app/schemas/calificaciones.py
from datetime import date
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

TipoRubro = Literal["Examen", "Tarea", "Cotidiano"]


class GrupoMini(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id_grupo: int
    name_grupo: str


class ProfesorMini(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id_profesor: int
    name_profesor: str
    sec_name_profesor: str


class AsignaturaMini(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id_asignatura: int
    name_asignatura: str


class AsignacionMini(BaseModel):
    """Profesor + materia de la asignacion a la que pertenece la evaluacion."""

    model_config = ConfigDict(from_attributes=True)

    id_profesor_asignatura_grupo: int
    profesor: ProfesorMini
    asignatura: AsignaturaMini


class EvaluacionBase(BaseModel):
    name_evaluacion: str = Field(min_length=1, max_length=100)
    tipo: TipoRubro = "Examen"
    periodo: int = Field(ge=1, le=4)
    porcentaje: float = Field(ge=0, le=100)
    fecha: date | None = None


class EvaluacionCreate(EvaluacionBase):
    # Se acepta la asignacion (nuevo modelo) o el grupo directo (compatibilidad).
    # Si viene la asignacion, el grupo se deriva de ella en el endpoint.
    id_grupo: int | None = None
    id_profesor_asignatura_grupo: int | None = None

    @model_validator(mode="after")
    def _requiere_grupo_o_asignacion(self) -> "EvaluacionCreate":
        if self.id_grupo is None and self.id_profesor_asignatura_grupo is None:
            raise ValueError(
                "Se requiere id_grupo o id_profesor_asignatura_grupo"
            )
        return self


class EvaluacionUpdate(BaseModel):
    name_evaluacion: str | None = Field(default=None, min_length=1, max_length=100)
    tipo: TipoRubro | None = None
    periodo: int | None = Field(default=None, ge=1, le=4)
    porcentaje: float | None = Field(default=None, ge=0, le=100)
    fecha: date | None = None
    id_grupo: int | None = None
    id_profesor_asignatura_grupo: int | None = None
    activo: bool | None = None


class EvaluacionOut(EvaluacionBase):
    model_config = ConfigDict(from_attributes=True)

    id_evaluacion: int
    id_grupo: int
    activo: bool
    grupo: GrupoMini
    id_profesor_asignatura_grupo: int | None = None
    asignacion: AsignacionMini | None = None


class NotaRegistroIn(BaseModel):
    id_estudiante: int
    valor: float | None = Field(default=None, ge=0, le=100)


class NotasBatchIn(BaseModel):
    registros: list[NotaRegistroIn]


class NotaRosterItem(BaseModel):
    id_estudiante: int
    name_estudiante: str
    sec_name_estudiante: str
    valor: float | None = None


class NotasRosterOut(BaseModel):
    id_evaluacion: int
    name_evaluacion: str
    registros: list[NotaRosterItem]
